lms/lmsdb/models.py

```diff
@@ -304,15 +354,38 @@ class Exercise(BaseModel):
     due_date = DateTimeField(null=True)
     notebook_num = IntegerField(default=0)
     order = IntegerField(default=0, index=True)
+    course = ForeignKeyField(Course, backref='exercise')
+    number = IntegerField()
+
+    class Meta:
+        indexes = (
+            (('course_id', 'number'), True),
+        )
 
     def open_for_new_solutions(self) -> bool:
         if self.due_date is None:
             return not self.is_archived
         return datetime.now() < self.due_date and not self.is_archived
 
     @classmethod
-    def get_objects(cls, fetch_archived: bool = False):
-        exercises = cls.select().order_by(Exercise.order)
+    def get_objects(
+        cls, user_id: int, fetch_archived: bool = False,
+        from_all_courses: bool = False,
+    ):
+        user = User.get(User.id == user_id)
+        exercises = (
+            cls
+            .select()
+            .join(Course)
+            .join(UserCourse)
+            .where(UserCourse.user == user_id)
+            .switch()
+            .order_by(UserCourse.date, Exercise.number, Exercise.order)
+        )
+        if not from_all_courses:
+            exercises = exercises.where(
+                UserCourse.course == user.last_course_viewed,
```

Comment:
`user.last_course_viewed` is nullable, we should first check it's not null (even though there is currently no flow to call `get_objects` without choosing a course, let's be a bit defensive)